### ISSUE TITLE
[RFC] - Serverless prerendering, partial hydration, and ESM bundles

### ISSUE THREAD

**Issue by tannerlinsley** [react-static/react-static#1180]
To drive React Static to be even more performant, we need to have a discussion around these three things:

- Serverless prerendering
- Partial hydration
- ESM bundles

Each of these technologies would make React Static sites more performant and also ship less code to the user or incrementally load it only when they need it. To do this, we need to:

- Learn more about these technologies and become fluent in their concepts
- Plan their implementation
- Execute

If you have *meaningful* input, ideas, or thoughts on these subjects, please comment below. If you approve of this roadmap, give it an emoji!

**Comment by sw-yx** [comment]
ESM bundles seems easiest. partial hydration doesnt seem necessary with static files? i could be misunderstanding. for serverless rendering, some coordination with the platform will be needed. from the netlify side, i dont even know how to support this.

**Comment by tannerlinsley** [comment]
Yeah, partial hydration is more common with isomorphic and streaming, but at the same time, the future could be that parts of the DOM remain static as separate feature-level chunks flow in asynchronously and partially hydrate different parts of the dom. Those hydration events could even be triggered by user interaction of sorts.

**Comment by tannerlinsley** [comment]
On the topic of SPR, Next.js is poised to do this best, since it's built to run as server-less functions. Part of me wonders if this is where static sites are going TBH. I think it's pretty exciting. Talking about incremental builds is one thing, but truly dynamic-static builds is on another level. Clearly, this would also take reasoning about your site on the route-level to another level as well. I know for any static sites that rely on medium-to-high data update frequency, building the whole site can be a pain.

Let's just explore the implications of what this could mean for React Static. It may not change anything at all.

**Comment by bram-l** [comment]
Partial hydration sounds brilliant. It's something I've been thinking about, but I never really came up with a good solution for it. Some transpilation step seems to be required to replace the static components. Or simply a wrapper component that is loaded from a module that can be conditionally swapped during the Webpack build? I'm curious to hear what are your ideas on this!

As for the (serverless) prerendering, it seems a whole different game to me. The beauty of static site generation is that it's just simple stupid static files coming out, which can be hosted anywhere, with low costs, high performance, and high scalability out of the box. Plus reliability, there are no more moving parts once it's built (apart from the magic happening client side of course). I feel like dynamic prerendering would take away some of these advantages. Also to reach a similar performance some caching would be required, and the whole thing becomes much more complex. In general I'd prefer to use a framework with some limitations, but that's just doing what it's doing best, without endless options. When a project requires more flexibility, for example in dealing with user data, I might just need a different framework.

**Comment by bram-l** [comment]
When it comes to the roadmap of this project, I think it would be a huge win to have (more) tests, both unit tests and end-to-end tests. When I was working on some bug fixes I noticed it can be really tricky to test for regressions, since there are many different contexts in which the code can be executed, with data coming from different sources (server side or client side, dynamic dev server or static production build, embedded data or remote data, preloaded data or non preloaded data). To be confident that a certain fix/change works in all those different situations automated testing would be extremely helpful!

**Comment by mikestopcontinues** [comment]
I've been thinking a lot about this recently. Haven't found any path convincing enough yet to explore, but I'll describe my use case:

I have a generic react app that individual subscribers customize given their theming, content, and subscription tier. These sites need to be delivered to my subscribers end-users worldwide, so they need to be fast. Additionally, I need to be able to run multiple simultaneous tests to optimize these sites around conversion.

So I've got three tiers of data that needs to be folded into my app before end-users see it:
* Core data: feature flags, config, multi armed bandit test variants etc.
* Subscriber data: Content, theme
* Multi Armed Bandit Test selections: Selection for this user need to happen near to page-load, but content flash needs to be prevented

And some data that can wait until end-users load a page:
* User data: login status, etc. (depending on implementation, I could pull this in with the test data. E.g. in a post-cache request lambda@edge)
* Dynamic content: comment threads, etc. (though if pre-rendered, I could inject a flat version and check for updates after load)

My out-of-the-box options:
* Basic Static (React-Static): After every change event, statically re-render the subscriber's app. Great for first- and non-first-requests, but how about for running tests? It means rendering a version of the app for every permutation of tests. Doable, but uggh.
* Basic SSR (Razzle): Compile code up front, then pull in all data during origin requests, and cache so that end-users never know how long it really takes to render the first-page. Code-splitting is an after-thought, so it's not going to client-nav much. Also, there's lots of permutations, so the CDN hit-rate will be low.
* Code-split SSR (Next.js): Next adds great code-splitting to SSR solutions. Origin-requests are still heavy, but non-first-requests are better. On the other hand, Next doesn't static-cache resources for the routes, which leaves it leagues behind react-static in client-side navigation speed. Plus, there's still the hit-rate problem.

What I'd really like is rendering stages, where I can pre-build an app so that very little additional data needs to be added at request-time. For instance, 
1. At compile time, store test variant data with the code-split bundles.
2. On subscriber data change events, static render a base version of the site (for bots, for admin/editor viewers, and low-performance devices)
3. At request time, after the base version is retrieved from the cache/origin, SSR hydrate and re-render given this particular permutation of test cases. Cache that version (for future end users with the same test cases) and send to the end user. (The code for this step can leverage Lambda@Edge or related solutions in other clouds.)

React-Static specializes in a fusion of step 1 and 2. And given that in my use case, it's a background process, that's not so bad. That said, I see steps 2-3 where the real magic of this new pattern happens. To make re-rendering really shine, you'd want to be able to build/rebuild on entire site or individual routes. So the pattern is something like:
1. BUILD: Compiles code, exports static data, splits code
2. REBUILD [path]: Re-export static data for matching paths. Overwrite routeData and staticData where the REBUILD provides new data. Skip compilation and code splitting, since it's already done. Perhaps it skips any routes where the routeData serializes the same as before.

This is turned out way longer than expected. As it stands, I'm still interested to see if I can finagle react-static to help meet my use-case, as I really don't like working with Next, but it seems like my best option. 

In any case, I hope this helps flesh out the feature, because I would love to see react-static take the lead on this.

**Comment by eyelidlessness** [comment]
If this is still being tracked, I’ll add some thoughts:

- This is probably a more valuable feature than people realize. Lots of sites, even heavily interactive apps, ship a ton of static junk in the bundle. Sometimes significantly more than the actually interactive components and their dependencies. This isn’t just a page weight issue, it also means static nodes are redundantly re-rendered/often blocked on TTI for basically no purpose.
- There are two ways existing solutions address this:
  - special opt-in/opt-out APIs designate what should be included in the bundle and how it’s hydrated; there are a bunch of prototypes taking this approach, and some existing/early stage solutions like Microsite and Astro.
  - identifying interactive components during compilation; as far as I know currently only Marko offers this in the JS/TS ecosystem.
  
  Both approaches have pros and cons. I originally thought I wanted the compiler/Marko approach, who wants to remember to write more code, after all? I ultimately changed my mind, preferring opt-in APIs because they provide more visibility into what you’re bringing into the client bundle.
  - React Server Components is likely to introduce a third approach: file naming convention. This will have a similar DX to the first approach.
- Many of the current solutions depend on particular build configurations that determine what to bundle based on the AST, but there’s potentially a simpler (JSX-specific, so out of scope for Astro) solution in a custom JSX pragma, marking static components as noops. This would have an added benefit by allowing those islands to share context and interop like they’re on a normal, full page app.
- React/compatible approaches typically mount separate trees under a containing `div`. Please, **please** don’t do this. One of the compelling use cases for partial hydration is to progressively enhance HTML developed with the same component model as the client bundle. Wrapper divs break a ton of CSS fallbacks that depend on sibling selectors. The way Microsite solves this is by rendering comment boundaries around would-be `Fragment`s during build. This works better because it matches the DOM structure defined by developers.
